### ISSUE TITLE
MUIA_TextEditor_Rows shouldn't affect MaxHeight

### ISSUE THREAD

**Issue by afalkenhahn** [amiga-mui/texteditor#36]
I'm trying to change TextEditor.mcc's default height of 15 rows to something smaller by setting `MUIA_TextEditor_Rows`. However, this has the effect that TextEditor.mcc will also set the maximum height to what has been passed to `MUIA_TextEditor_Rows`, resulting in a gadget that's not vertically resizable. Could this be changed in a way that `MUIA_TextEditor_Rows` will just affect the minimum and default height of the TextEditor gadget but not the maximum height? IMHO this would make much more sense.

If you don't want to change the existing behaviour, maybe a new tag could be introduced.

**Comment by afalkenhahn** [comment]
Hello, is anybody out there?

**Comment by alpyre** [comment]
Hi. Actually this did not make much sense to me. If a text editor gadget has a set max rows and/or columns but the displayed gadget size could be bigger, that would create areas within the gadget which will be clickable but the cursor won't be allowed to go there. That would be terribly confusing to the user IMHO.

However, if that really is the behaviour you want to achieve, you can always create a texteditor gadget without a frame in a group that has string frame together with some rectangle objects. This shouldn't mandate a change in texteditor class' behaviour.

**Comment by afalkenhahn** [comment]
Well, actually, I just want to find a way how to change the gadget's default height to something less than 15 rows while still maintaining the gadget's vertical sizability. I don't see any way to do this. As I said, `MUIA_TextEditor_Rows` doesn't really help because it will fix the maximum size to the number of rows passed to it. So is there no way to change the gadget's default height while still maintaining vertical sizability?

**Comment by alpyre** [comment]
Oh. I think now I understand. I think you can achieve that by setting vertical weights to your gadgets (with MUIA_VertWeight).

**Comment by afalkenhahn** [comment]
Huh? What does the vertical weight have to do with my problem? Sorry, but I don't think you know what you're talking about.

The issue can clearly be seen in Dispatcher.c:

```
  if(data->Rows != 0)
  {
    LONG height = data->Rows * fontheight;

    mi->MinHeight += height;
    mi->DefHeight += height;
    mi->MaxHeight += height;
  }
  else
  {
    mi->MinHeight +=  1 * fontheight;
    mi->DefHeight += 15 * fontheight;
    mi->MaxHeight = MUI_MAXMAX;
  }
```

If `MUIA_TextEditor_Rows` is set, the height of the gadget will be completely fixed. If it's not specified, the height will be resizable but **NOT** smaller than the hardcoded 15 rows. That's why I think there's no way to solve my problem without touching the code. Are you the maintainer of TextEditor,mcc now? Where are Thore and Jens?

**Comment by jens-maus** [comment]
@softwarefailure 
> Where are Thore and Jens?

We are spending our precious easter time with our families and not with a defacto dead platform, I would say ;)

Nevertheless, TextEditor.mcc is fully open sourced on GitHub. So if you find something missing or broken, fix it yourself and submit a PullRequest and we will certainly integrate it ASAP.

**Comment by afalkenhahn** [comment]
> We are spending our precious easter time with our families 

Well, but my original request is from 15th March, so...

> Nevertheless, TextEditor.mcc is fully open sourced on GitHub. So if you find something missing or broken, fix it yourself and submit a PullRequest and we will certainly integrate it ASAP.

I could certainly come up with a patch but I think we should first agree on a solution on how to fix this. It's probably better to clarify that before writing the code.

**Comment by alpyre** [comment]
I'm very sorry for completely misunderstanding your issue. Now I've got it.

You can change the default height by subclassing MUIC_TextEditor and overriding MUIM_AskMinMax. Something like:
`
struct cl_Data
{
  ULONG Rows;
};

//Overridden OM_NEW
static ULONG m_New(struct IClass* cl, Object* obj, struct opSet* msg)
{
  struct cl_Data *data;
  ULONG rows = GetTagData(MUIA_TextEditor_Rows, 0, msg->ops_AttrList);

  if (obj = (Object *) DoSuperNew(cl, obj, TAG_MORE, msg->ops_AttrList))
  {
    data = (struct cl_Data *) INST_DATA(cl, obj);
    data->Rows = rows;

    return((ULONG) obj);
  }

return NULL;
}

//Overridden mAskMinMax()
static ULONG mAskMinMax(struct IClass *cl, Object *obj, struct MUIP_AskMinMax *msg)
{
  struct cl_Data *data = INST_DATA(cl, obj);
  struct MUI_MinMax *mi;
  LONG fontheight;

  DoSuperMethodA(cl, obj, (Msg)msg);

  mi = ((struct MUIP_AskMinMax *)msg)->MinMaxInfo;

  fontheight = _font(obj)->tf_YSize;

  if(data->Rows == 0)
  {
    mi->DefHeight -= 10 * fontheight; // This will set your DefHeight to 15 - 10 = 5
  }

  return 0;
}

//Dispatcher
static DISPATCHERPROTO(cl_Dispatcher)
{
  struct cl_Data *data;
  if (! (msg->MethodID == OM_NEW)) data = INST_DATA(cl, obj);

  switch(msg->MethodID)
  {
    case OM_NEW:
      return m_New(cl, obj, (struct opSet*) msg);
    case MUIM_AskMinMax:
      return mAskMinMax(cl, obj, (struct MUIP_AskMinMax*) msg);

    default:
      return DoSuperMethodA(cl, obj, msg);
  }
}
`

Yes this will be somewhat hacky but otherwise you will have to introduce a new (I..) attribute in TextEditor_MCC's code.

**Comment by afalkenhahn** [comment]
Yeah, of course overriding mAskMinMax with a custom implementation will solve the problem but of course I was looking for a TextEditor.mcc update which offers a way to do this using inhouse means but maybe the faster route is really through overriding mAskMinMax.

**Comment by tboeckel** [comment]
> Well, but my original request is from 15th March, so...

Well, opening a new issue is simply a request towards the developers to take care about the issue. There is no guarantee that it will be handled, nor a deadline when it will be handled. As Jens already pointed out, we have families, spare time and lots of other stuff to care about which are far more important than this project.

> > Nevertheless, TextEditor.mcc is fully open sourced on GitHub. So if you find something missing or broken, fix it yourself and submit a PullRequest and we will certainly integrate it ASAP.
> 
> I could certainly come up with a patch but I think we should first agree on a solution on how to fix this. It's probably better to clarify that before writing the code.

You still could prepare a pull request which we alltogether then can discuss about in which way it might be merged into the project. But simply stating "I found an issue and you MUST fix it immediately" is not going to work.

Anyway... I finally found the time to implement the two new attributes MUIA_TextEditor_FreeHoriz/Vert to override the maximum limits due to MUIA_TextEditor_Columns/Rows. Just set these to TRUE and the TE object will be fully resizeable again.

[MCC_TextEditor-15.57.zip](https://github.com/amiga-mui/texteditor/files/8530116/MCC_TextEditor-15.57.zip)
